backend/v3/instrumentation/otel.go

```diff
@@ -0,0 +1,133 @@
+package instrumentation
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/log/global"
+	"go.opentelemetry.io/otel/propagation"
+	"go.opentelemetry.io/otel/sdk/log"
+	"go.opentelemetry.io/otel/sdk/resource"
+	semconv "go.opentelemetry.io/otel/semconv/v1.17.0"
+
+	"github.com/zitadel/zitadel/cmd/build"
+)
+
+//go:generate enumer -type=ExporterType -trimprefix=ExporterType -text -linecomment
+type ExporterType int
+
+// ExporterType defines the type of exporter to use.
+// - metrics supports all types
+// - tracing supports all types except Prometheus
+// - logging supports all types except Google and Prometheus
+// - profiling supports only Google and None
+const (
+	// Empty line comment sets empty string of unspecified value
+	ExporterTypeUnspecified ExporterType = iota //
+	ExporterTypeNone
+	ExporterTypeStdOut
+	ExporterTypeStdErr
+	ExporterTypeGRPC
+	ExporterTypeHTTP
+	ExporterTypeGoogle
+	ExporterTypePrometheus
+)
+
+type ExporterConfig struct {
+	Type            ExporterType
+	Endpoint        string
+	Insecure        bool
+	BatchDuration   time.Duration
+	GoogleProjectID string // only for tracing, metrics and profiler
+}
+
+// setupOTelSDK bootstraps the OpenTelemetry pipeline.
+// If it does not return an error, make sure to call shutdown for proper cleanup.
```

Comment:
shouldn't shutdown always be called? if starting the last provider fails all the other ones would not be shut down gracefully

Reply:
That's what HandleErr is for. The function calls shutdown itself in case of an error. Typically callers of a go function only check for an non-nil error and assume remaining return values are nil.

```
https://github.com/zitadel/zitadel/blob/24cd627d8874923b10c7df8b88d65ca1b468568f/backend/v3/instrumentation/otel.go#L69-L71
```